[WIP] Typescript Typings

RXminuS opened the pull request:
There was a long discussion in https://github.com/ianstormtaylor/slate/issues/560 if we should include Typescript typings in the repo itself or not. The core team seamed geared towards keeping typings separate.

However, with recent big API changes the other attempts at external typings have almost fully broken. Personally I feel that the effort required to maintain a set of solid base typings while making a PR is much easier than a constant back and forth to an external repo. The auto-generated typings out there miss a lot of details (especially because of how changes are added to prototype classes)

So I've started to hand-type the entire library. My goal was to keep it as symbolically close to the source code as possible so that any contributers to the project should easily find what they are looking for. Sometimes it was actually super unclear from the source and docs what the code is supposed to do so I even think that types can help newcomers more easily figure out what is happening.

As a disclaimer, I'm very new to Typescript and am by no means an expert. Therefore there might be some patterns that can be polished a bit nicer, but am hoping for a bit of community help. My hope is that once these typings are done we can include them in the project. I'm even willing to commit myself to maintain them as part of PR's or donate some money for others to maintain them if it's really that big a hassle.

On a side-note: I'm also finding a few bugs because of the typings, I'll post fixes as separate PR's

- [x] Core Typings
- [ ] Packages
- [ ] Tests
- [ ] Build Pipeline

lichenglu commented:
@RXminuS Would love to help if you want to take something off your plate!

zhujinxuan commented:
Hi, I would suggest not to document `getBlocksAsArray` and other `...AsArray` functions (or docunmenting them as private).  Because these functions are internal functions to speed up functions like `getBlock`

RXminuS commented:
Thank you @zhujinxuan I've fixed that. Also added a bunch of missing methods and used the https://github.com/piotrwitek/utility-types package to not re-invent the wheel for DeepImmutable etc. Just need to cleanup and add a few test and then will push later this week. Also added typings for a few of the other side packages (html-serializer and react) @lichenglu if you want to help you could add typings for the other side-packages? They should be fairly straight forward and for inspiration you could have a look at the current ones on DefinitelyTyped?

zhujinxuan commented:
Hi, I am not sure whether this is doable in ts checking:
```
  const child = document.getNodeByKey(key);
  if (child) return;
  const parent :: (NotNULL)= document.getParentByKey(key);
  const path :: (NotNULL)= parent.getPath(key)
```

Sometimes, we want to assert non-null according to our knowledge, but I am not sure whether ts allow us to assert non-null.

ianstormtaylor commented:
Fixed by https://github.com/ianstormtaylor/slate/pull/3093.
